### DISCUSSION TITLE
VI support in editor pane

### DISCUSSION THREAD

**Discussion by masa300V** [asciidocfx/AsciidocFX#495] [Ideas]
Hello, thanks for this great tool and this is a feature request from a VI user.

Can we have an option to choose a full-featured terminal emulator at existing editor pane where we have ACE now?

In this terminal, if implemented, I want to run [Neovim](https://github.com/neovim/neovim).
And I think we can setup some simple / basic I/F between Neovim and AsciidocFX by utilizing Neovim's API for opening files, start docbook/pdf conversions.

I tried to use Neovim in the bottom "Pseudo Terminal Emulator" but not so successful.

**Comment by gillesB** [comment]
Depending on how many VI commands you need, you could try the VIM-mode of ACE?
I have not tried the built-in mode, except in the browser (https://ace.c9.io/build/kitchen-sink.html), where it behaves strangely (Ctrl+N opens a new browser window, Ctrl+R is Undo).
But perhaps it reacts better in a desktop application?

**Reply by masa300V** [reply]
Thank you very much!!!
I figured out that keybinding-vim.js does the job as you suggested.

**Reply by gillesB** [reply]
I finally tried it myself and to activate the VIM-mode I did the following:

* go to the AsciidocFX installation folder then to `./conf/public/js`
* open `editor.js`
* add the line `editor.setKeyboardHandler("ace/keyboard/vim");` below `editor.setTheme("ace/theme/xcode");`
* (re)start AsciidocFX

I do not know if this is the best way to use VIM-mode but it seems to work.
Furthermore I think the modification will be overwritten on the next AsciidocFX update.

I also managed to configure autocompletion (Ctrl+Space, see also [ext-modelist.js](https://stackoverflow.com/a/19730470/2021763)) and relative line numbers:

* in `editor.js` add `ace.require("ace/ext/language_tools");` as first line
* change the options in `editor.js` as follows

```
editor.setOptions({
    enableSnippets: true,
    dragEnabled: true,
    enableBasicAutocompletion: true,
    relativeLineNumbers:true
});
```

* finally in `public/editor.html` add `<script src="/afx/resource/ace/src/?p=ace/src/ext-language_tools.js" type="text/javascript" charset="utf-8"></script>` after the line with `ext-modelist.js`

**Comment by masa300V** [comment]
Today, I switched to 1.8.6, I'm using Windows version.
I set my "terminal" shell to git-bash and I see I can start Neovim in the terminal via the git-bash.
With a color availability, I switched my colorscheme of less-colorful one and working.

Just an update for people likes vi ...

**Comment by jlanzarotta** [comment]
I tried adding `editor.setKeyboardHandler("ace/keyboard/vim");`, as mentioned above. However; this does not seem to work, at least for me.  Any suggestions on how to get Vi(m) type keybindings working?  I do see a file named _keybind-vim.js_; however, that does not seem to work either.

Thanks.
